### PR TITLE
update cse and cloud-init data for containerd [2/4]

### PR DESCRIPTION
**Pull request by eriksywu**
cse_main.sh:
- install crictl
- move order of install around for containerruntime

cse_config.sh/cse_util.sh:
- minor go-template updates

cse_install.sh:
- refactor cleanUpKubeProxyImages and cleanUpHyperkubeImages

parts/linux/cloud-init/nodecustomdata.yml:
- update containerd/config.toml to version = 2 and add additional configs
- [bug fix] /etc/sysctl.d/11-containerd.conf should be applied to all containerd scenarios

### CODE REVIEW

File: parts/linux/cloud-init/artifacts/cse_main.sh [hunk 1]
```diff
@@ -45,10 +45,6 @@ fi
 
 configureAdminUser
 
-{{- if not NeedsContainerd}}
-cleanUpContainerd
-{{end}}
-
```
Comment (xuto2) [inline parts/linux/cloud-init/artifacts/cse_main.sh:51]: I think we still need this for the previous VHDs with  containerd and dockerd

File: pkg/agent/params.go [hunk 2]
```diff
@@ -208,6 +213,11 @@ func assignKubernetesParameters(properties *datamodel.Properties, parametersMap
 			addValue(parametersMap, "networkPlugin", kubernetesConfig.NetworkPlugin)
 			addValue(parametersMap, "networkMode", kubernetesConfig.NetworkMode)
 			addValue(parametersMap, "containerRuntime", kubernetesConfig.ContainerRuntime)
+			if kubernetesConfig.ContainerRuntime == "containerd" {
+				addValue(parametersMap, "cliTool", "ctr")
+			} else {
+				addValue(parametersMap, "cliTool", "docker")
+			}
```
Comment (xuto2) [inline pkg/agent/params.go:220]: It's a bit confusing to see two places assigning the same parameters. Can we just do it from agent pool profile? I kind of imagine the agent pool profile should be updated in RP side already, if it needs to inherit any property from cluster.

Not a blocking issue but worth consideration , along with the teleport enablement flag as we discussed in the design review - historically container runtime is a cluster level setting as we don't have multi agent pool, but now, most of the setting becomes agent pool specific, e.g os config, container run time, k8s version, etc. Long term, I think it's a better design to completely move them from cluster to agent pool. The cluster level setting is just good for control plane side components. WDYT?

File: pkg/agent/testdata/AKSUbuntu1604+CustomKubeletConfig+CustomLinuxOSConfig/line16.sh [hunk 3]
```diff
@@ -86,13 +84,17 @@ fi
...
 configureK8s
 
 configureCNI
 
 
+
+
+ensureDocker
+
+
```
Comment (xuto2) [inline pkg/agent/testdata/AKSUbuntu1604+CustomKubeletConfig+CustomLinuxOSConfig/line16.sh:96]: nit: Too many space lines, can we try to remove them?

File: parts/linux/cloud-init/artifacts/cse_install.sh [hunk 4]
```diff
@@ -310,41 +309,48 @@ removeContainerImage() {
...
     }
-    export -f cleanUpHyperkubeImagesRun
-    retrycmd_if_failure 10 5 120 bash -c cleanUpHyperkubeImagesRun
-    echo $(date),$(hostname), endCleanUpHyperkubeImages
+    export -f cleanupAllImagesRun
+    retrycmd_if_failure 10 5 120 bash -c cleanupAllImagesRun
+}
+
+cleanUpHyperkubeImages() {
+    echo $(date),$(hostname), startCleanUpKubeProxyImages
```
Comment (xuto2) [inline parts/linux/cloud-init/artifacts/cse_install.sh:339]: s/KubeProxy/Hyperkube

File: parts/linux/cloud-init/artifacts/cse_install.sh [hunk 5]
```diff
@@ -310,41 +309,48 @@ removeContainerImage() {
     fi
 }
 
-cleanUpHyperkubeImages() {
-    echo $(date),$(hostname), startCleanUpHyperkubeImages
-    function cleanUpHyperkubeImagesRun() {
-        images_to_delete=$(docker images --format '{{OpenBraces}}.Repository{{CloseBraces}}:{{OpenBraces}}.Tag{{CloseBraces}}' | grep -vE "${KUBERNETES_VERSION}$|${KUBERNETES_VERSION}.[0-9]+$|${KUBERNETES_VERSION}-|${KUBERNETES_VERSION}_" | grep 'hyperkube')
+cleanUpAllImages() {
```
Comment (xuto2) [inline parts/linux/cloud-init/artifacts/cse_install.sh:312]: cleanUpImages?  It's actually cleaning any image, but not all

**Review by xuto2 — CHANGES_REQUESTED** [review]
once this is merged pls vendor to AKS RP to give it a try

File: parts/linux/cloud-init/artifacts/cse_main.sh (hunk 1 above)
Comment (eriksywu) [reply parts/linux/cloud-init/artifacts/cse_main.sh:51]: added back in

File: pkg/agent/params.go (hunk 2 above)
Comment (eriksywu) [reply pkg/agent/params.go:220]: Agreed. Will remove the addition here. We should let aks-rp handle all the agentpool profile configurations. 

We can probably also remove the line above where it sets the runtime but I will leave it in for now, to be safe. At GA aks-rp will set the appropriate agentpool-level runtime configuration based on the VHD distro so this should def just be an agentpool-level config here as well.

I'll make an ADO item to track this small cleanup task post-GA.

File: parts/linux/cloud-init/artifacts/cse_install.sh (hunk 5 above)
Comment (eriksywu) [reply parts/linux/cloud-init/artifacts/cse_install.sh:312]: renaming o cleanupImages

**Comment by eriksywu** [comment]
passed check-in e2e suite and manual private rp e2e (with dynamic-kubelet-config hotfix)

File: pkg/agent/testdata/AKSUbuntu1604+CustomKubeletConfig+CustomLinuxOSConfig/line16.sh [hunk 6]
```diff
@@ -86,13 +84,17 @@ fi
...
 
-ensureContainerRuntime
-
 configureK8s
 
 configureCNI
 
 
+
+
```
Comment (xuto2) [inline pkg/agent/testdata/AKSUbuntu1604+CustomKubeletConfig+CustomLinuxOSConfig/line16.sh]: still unnecessary empty line, but we can fix it in the next PR.

File: pkg/agent/testdata/AKSUbuntu1604+CustomKubeletConfig+CustomLinuxOSConfig/line23.sh [hunk 7]
```diff
@@ -259,41 +258,44 @@ removeContainerImage() {
...
+                
+            done
         fi
     }
-    export -f cleanUpHyperkubeImagesRun
-    retrycmd_if_failure 10 5 120 bash -c cleanUpHyperkubeImagesRun
-    echo $(date),$(hostname), endCleanUpHyperkubeImages
+    export -f cleanupAllImagesRun
+    retrycmd_if_failure 10 5 120 bash -c cleanupAllImagesRun
+}
```
Comment (xuto2) [inline pkg/agent/testdata/AKSUbuntu1604+CustomKubeletConfig+CustomLinuxOSConfig/line23.sh]: let's remove the unnecessary empty lines in a separate PR if you need to merge this now.

File: pkg/agent/testdata/AKSUbuntu1604+CustomKubeletConfig+CustomLinuxOSConfig/line23.sh [hunk 8]
```diff
@@ -259,41 +258,44 @@ removeContainerImage() {
     fi
 }
 
-cleanUpHyperkubeImages() {
-    echo $(date),$(hostname), startCleanUpHyperkubeImages
-    function cleanUpHyperkubeImagesRun() {
-        images_to_delete=$(docker images --format '{{.Repository}}:{{.Tag}}' | grep -vE "${KUBERNETES_VERSION}$|${KUBERNETES_VERSION}.[0-9]+$|${KUBERNETES_VERSION}-|${KUBERNETES_VERSION}_" | grep 'hyperkube')
+cleanUpAllImages() {
```
Comment (xuto2) [inline pkg/agent/testdata/AKSUbuntu1604+CustomKubeletConfig+CustomLinuxOSConfig/line23.sh]: Did you forget to send a new commit? This is still not renamed.

File: parts/linux/cloud-init/artifacts/cse_install.sh (hunk 5 above)
Comment (xuto2) [reply parts/linux/cloud-init/artifacts/cse_install.sh]: still not resolved.

**Review by xuto2 — APPROVED** [review]

File: pkg/agent/testdata/AKSUbuntu1604+CustomKubeletConfig+CustomLinuxOSConfig/line23.sh (hunk 8 above)
Comment (xuto2) [reply pkg/agent/testdata/AKSUbuntu1604+CustomKubeletConfig+CustomLinuxOSConfig/line23.sh]: nvm. I looked at the wrong iteration.

File: parts/linux/cloud-init/artifacts/cse_install.sh (hunk 5 above)
Comment (xuto2) [reply parts/linux/cloud-init/artifacts/cse_install.sh]: nvm. Looked at the wrong iteration.

**Pull request merged**